Embedded Software/TelemetryControlUnit/src/logger.cpp

```diff
@@ -0,0 +1,250 @@
+  message_format_t *messageBuf = usingBuf1 ? messageBuf1 : messageBuf2; 
+  int bufLength = usingBuf1 ? buf1Length : buf2Length;
+
+  // only write on certain conditions
+  if ((millis() - lastLogTime) <= minLogFrequency && bufLength < MAX_BUFFERED_MESSAGES) {
+      return LOGGER_ERROR_NO_WRITE;
+  }
+
+  usingBuf1 = !usingBuf1; // Switch main log buffer during write process
+
+  // create a new file if current one is too large
+  if (logFile.size() > MAX_LOG_FILE_SIZE) {
+    logFile.close();
+    setFileName();
+    logFile = SD.open(fileName, FILE_WRITE);
+  }
+
+  DPRINT(F(" -- "));
+  DPRINT(F("Added from buf "));
+  if (usingBuf1) {
+    DPRINT(F("2: "));
+  } else {
+    DPRINT(F("1: "));
+  }
+  DPRINT(bufLength);
+  DPRINT(F(" -- \n"));
+
+  if (logFile) {
+    DPRINTLN(F("Writing to SD card..."));
+
+    // write all buffered messages to the SD card
+    for (int i = 0; i < bufLength; i++) {
+      int writeLength = logFile.print(messageBuf[i].timestamp);
+      // checks for error on write (assumes rest are fine if this passes)
+      if (writeLength == 0) {
+        reset();
+        return LOGGER_ERROR_SD_CARD;
+      }
+      logFile.print(F(" "));
+      logFile.print(messageBuf[i].id);
+      logFile.print(F(" "));
+      logFile.print(messageBuf[i].length);
+      logFile.print(F(" ["));
+      for (int j = 0; j < messageBuf[i].length; j++) {
+        logFile.print(messageBuf[i].dataBuf[j]);
+        if (j != messageBuf[i].length - 1) {
+          logFile.print(F(","));
+        }
+      }
+      logFile.print(F("]\n"));
+    }
+
+    lastLogTime = millis();
+    
+    // clear bufLength variables (must clear the opposite of whatever one is in use)
+    if (usingBuf1) {
+      DPRINTLN(F("Reset buffer 2"));
+      buf2Length = 0;
+    } else {
+      DPRINTLN(F("Reset buffer 1"));
+      buf1Length = 0;
+    }
+    logFile.flush();
+    return LOGGER_SUCCESS;
+  } 
+  else {
+    DPRINTLN(F("Could not open file on SD card"));
+    reset();
+    return LOGGER_ERROR_SD_CARD;
+  }
+}
+
+
+/**
+ * @brief Adds the given data plus a generated timestamp to the message buffer.
+ *        Uses a time since startup if the RTC is not in use, or the real time otherwise.
+ * 
+ * @param id  Message id
+ * @param len Message length
+ * @param buf Data bytes (Array of length 'len')
+ */
+int LoggerBufferMessage(uint32_t id, uint8_t len, const uint8_t *buf) {
+  if (!initialized) {
+    return LOGGER_ERROR_SD_CARD;
+  }
+
+  noInterrupts();
+  message_format_t *messageBuf = usingBuf1 ? messageBuf1 : messageBuf2; 
+  int bufLength = usingBuf1 ? buf1Length : buf2Length;
+
+  if (bufLength >= MAX_BUFFERED_MESSAGES) {
+    DPRINTLN(F("Tried to write to already full buffer"));
+    return LOGGER_ERROR_BUFFER_FULL;
+  }
+
+  // add data to message buffer
+  getTimestamp(messageBuf[bufLength].timestamp);
+  messageBuf[bufLength].id = id;
+  messageBuf[bufLength].length = len;
+  memcpy(messageBuf[bufLength].dataBuf, buf, len);
```

Comment:
You sure we need a memcpy here? imo memcpy is kinda dangerous if you don't use it right and could just be substituted for a for loop. memcpy is cool tho so as long as you're sure we won't get weird lengths and end up copying too much or too little data

Reply:
The dataBuf has space for 8 bytes and the max len is 8, so there will never be any overwriting. Could use a for loop but I wanted this operation to be as fast as possible as it's called from the CAN receive callback and interrupt handlers should be fast. Might add up if we're logging hundreds of messages per second.